pubsub/google/cloud/pubsub_v1/publisher/client.py

```diff
@@ -0,0 +1,162 @@
+            if batch is None or not batch.will_accept(message):
+                if not create:
+                    return None
+                batch = self._batch_class(
+                    autocommit=autocommit,
+                    client=self,
+                    settings=self.batch_settings,
+                    topic=topic,
+                )
+                self._batches[topic] = batch
+
+        # Simply return the appropriate batch.
+        return batch
+
+    def publish(self, topic, data, **attrs):
+        """Publish a single message.
+
+        .. note::
+            Messages in Pub/Sub are blobs of bytes. They are *binary* data,
+            not text. You must send data as a bytestring
+            (``bytes`` in Python 3; ``str`` in Python 2), and this library
+            will raise an exception if you send a text string.
+
+            The reason that this is so important (and why we do not try to
+            coerce for you) is because Pub/Sub is also platform independent
+            and there is no way to know how to decode messages properly on
+            the other side; therefore, encoding and decoding is a required
+            exercise for the developer.
+
+        Add the given message to this object; this will cause it to be
+        published once the batch either has enough messages or a sufficient
+        period of time has elapsed.
+
+        Example:
+            >>> from google.cloud.pubsub_v1 import publisher_client
+            >>> client = publisher_client.PublisherClient()
+            >>> topic = client.topic_path('[PROJECT]', '[TOPIC]')
+            >>> data = b'The rain in Wales falls mainly on the snails.'
+            >>> response = client.publish(topic, data, username='guido')
+
+        Args:
+            topic (~.pubsub_v1.types.Topic): The topic to publish
+                messages to.
+            data (bytes): A bytestring representing the message body. This
+                must be a bytestring.
+            attrs (Mapping[str, str]): A dictionary of attributes to be
+                sent as metadata. (These may be text strings or byte strings.)
+
+        Returns:
+            ~.pubsub_v1.publisher.futures.Future: An object conforming
+                to the ``concurrent.futures.Future`` interface.
+        """
+        # Sanity check: Is the data being sent as a bytestring?
+        # If it is literally anything else, complain loudly about it.
+        if not isinstance(data, six.binary_type):
+            raise TypeError('Data being published to Pub/Sub must be sent '
+                            'as a bytestring.')
+
+        # Coerce all attributes to text strings.
+        for k, v in copy.copy(attrs).items():
+            if isinstance(v, six.text_type):
+                continue
+            if isinstance(v, six.binary_type):
+                attrs[k] = v.decode('utf-8')
+                continue
+            raise TypeError('All attributes being published to Pub/Sub must '
```

Comment:
I think I prefer the way it is written over `if`/`elif`/`elif`, although I am struggling to come up with a reason why.